### PR TITLE
optimized spatialization

### PR DESCRIPTION
**Pull request by beasteers**
This refactors the spatialization code so that it is as vectorized as I could make it. I also added some indexing that reduced zero computations in the matrix mults.

To run the tests which currently compare end-to-end audio equality, do: `pytest tests`

you can listen to them here: `tests/output/generation`

Changes:
 * `spatialize` function
   * improve vectorization, interpretability, and reduce array copies
   * move to `spatialize.py`
   * encapsulate `scipy.signal.convolve` for single IRs
   * accept `irs=(n_ch, n_irs, n_ir_samples)` instead of `irs=(n_ir_samples, n_ch, n_irs)`
   * change argument name `s -> snr`
   * outputs identical to outputs from before changes

 * `core.py`
   * change `spatialize` import to `spatialize.py`
   * add progress bar to `synthesize_events_and_labels`
   * single IR handling removed and moved to `spatialize`
 * `utils.py`
   * added `set_seed` for test reproducibility
   * `spatialize` can be deleted **
 * `tests` (I considered making this a separate PR since this is just general regression testing)
   * `test_generate.py`: basic end-to-end generation tests. compares generated audio with trusted reference outputs
   * `data/*`: reference wav and csv files using original spatialize function
 * `tests/misc`: test utils
   * `profile_generation.py`: basically `example_generation.py` but with profiling
   * `profile_spatialize.py`: specifically profile `spatialize` over different input parameters
   * `compare_audio.py`: diff spectrograms of two audio files

Before merging, we can remove the following functions from `utils.py`:
 * `spatialize`
 * `_spatialize`
 * `stft_ham`
 * `ctf_ltv_direct`

Since there are test audio files in the commit history, make sure to squash the commits when merging.

### Benchmarking
Here's timings when generating 15 60s soundscapes with moving sources and a fixed random seed.

script: `tests/misc/profile_spatialize.py` (to use the old spatialization, comment out `from .spatialize import spatialize` from `core.py`)

Improvement: 610s (40.6s per audio file) -> 134s (8.9s per audio file).

Before:
image

After:
image

I isolated the spatialize function and systematically tested it using all random noise:
`sr = 24000, ir_samples = 7200`
image

### CODE REVIEW

**Comment by adrianSRoman** [comment]
This is awesome @beasteers! Implementation wise things look in great shape to me. Do you need help performing validation? I believe generating test sets and bench-marking with a pre-trained SELDnet would be a good sanity check. I see you also added some functional tests.

**Comment by beasteers** [comment]
Yeah @adrianSRoman if there are any other tests you'd like to run (e.g. validating with SELDnet), I'd love the help as I'm not familiar with the pipeline atm.

Also, if you could just reproduce the tests (`pytest tests`) on your end that would be great! soundata threw some zip errors when I was downloading the datasets so I don't fully trust my download.

If the tests fail (because there's something wrong with my datasets), you can try regenerating the test files with the old spatialization code by commenting out `from .spatialize import spatialize` in `core.py` and rerunning `pytest tests` . 

Then once the old spatialization files are generated, make those the new reference files
```bash
rm -r tests/data/generation
mv tests/output/generation tests/data/generation
```
and finally, uncomment `from .spatialize import spatialize` and rerun `pytest tests`. If they are different, then perhaps you could commit your corrected versions. Though I'm thinking that perhaps the better solution is to not make the tests dependent on dataset files (e.g. using artificial sources/IRs).

The functional test is based off how scaper does their tests, but lmk if you have other ideas.
